src/key_io.cpp

```diff
@@ -59,7 +64,20 @@ class DestinationEncoder
         std::vector<unsigned char> data = {1};
         data.reserve(53);
         ConvertBits<8, 5, true>([&](unsigned char c) { data.push_back(c); }, tap.begin(), tap.end());
-        return bech32::Encode(bech32::Encoding::BECH32M, m_params.Bech32HRP(), data);
+
+        std::string hrp = m_params.Bech32HRP();
+
+        if (m_silent_payment) {
+            hrp = m_params.SilentPaymentHRP();
+
+            std::ostringstream ss;
+            ss << std::setw(2) << std::setfill('0') << m_silent_payment_index;
```

Comment:
Another excellent suggestion @aureleoules !

I've been working on this.
Some examples:
`sprt1la68kq44e29c0p85rknf59qs5eww6vr3dw08arakura622ce7wcvygce0uqnw3q4` for identifier `8090879`.
`sprt1yvpttj5ts7z0g8dxng2ppfjua5c8z6u7068mdc8m5543nuascg33jlcpe5xc6` for identifier `35`.

Reply:
Great, how exactly is the identifier encoded in your examples?

Reply:
Currently the format is: `HRP + 00 (two digits) + V1 Segwit address (bech32m)`

But there is no need to be a segwit address. This is because in the first version the sender could pay for the original segwit address or tweak it to the silent address, but early discussions in this PR indicated a preference on the part of reviewers for a specific address for silent payments.

So the format in the examples above is different, which is:

`HRP + identifier (variable length) + compressed public key (both identifier and PK are in bech32m format)`

I haven't pushed it yet as the above format is not compatible with current code and also adds complexity to the code as it is not a valid scriptPubKey.

However, here is the code used to obtain the format above (in `key_io.cpp`):

<details><summary>Code</summary>
<p>

```c++
std::string EncodeSilentDestination(const CPubKey pubkey, const int32_t silent_payment_index)
{
    // The data_in is index + public key
    std::vector<unsigned char> data_in = {};
    std::vector<unsigned char> data_out = {};

    std::vector<unsigned char> index_data = {};
    unsigned char index_chunk = silent_payment_index & 0xFF;
    int8_t shift = 8;

    while(index_chunk != 0) {
        index_data.push_back(index_chunk);
        index_chunk = (silent_payment_index >> shift) & 0xFF;
        shift += 8;
    }

    data_in.insert(data_in.end(), index_data.begin(), index_data.end());
    data_in.insert(data_in.end(), pubkey.begin(), pubkey.end());

    ConvertBits<8, 5, true>([&](unsigned char c) { data_out.push_back(c); }, data_in.begin(), data_in.end());

    std::string hrp = Params().SilentPaymentHRP();

    return bech32::Encode(bech32::Encoding::BECH32M, hrp, data_out);
}

std::tuple<CPubKey, int32_t> DecodeSilentData(const std::vector<unsigned char>& data)
{
    if (data.size() <= 33) {
        return {CPubKey(), 0};
    }

    std::vector<unsigned char> pubkey_data(data.end() - 33, data.end());
    CPubKey pubkey{pubkey_data.begin(), pubkey_data.end()};

    std::vector<unsigned char> index_data(data.begin(), data.end() - 33);

    int32_t index = 0;
    int shift = 0;

    for (auto it = index_data.begin(); it != index_data.end(); it++) {
        index |=  (*it << shift);
        shift += 8;
    }

    return {pubkey, index};
}

std::vector<unsigned char> DecodeSilentAddress(const std::string& str)
{
    auto params{Params()};

    const auto& silent_payment_hrp = params.SilentPaymentHRP();
    auto dest_silent_payment_hrp = ToLower(std::string_view(str).substr(0, params.SilentPaymentHRP().size()));

    bool is_sp = dest_silent_payment_hrp == silent_payment_hrp;

    if (!is_sp) {
        return {};
    }

    std::vector<unsigned char> data;
    data.clear();
    const auto dec = bech32::Decode(str);
    auto dec_silent_payment_hrp = dec.hrp.substr(0, params.SilentPaymentHRP().size());

    if (dec.encoding != bech32::Encoding::BECH32M || dec.data.empty() || dec.hrp != silent_payment_hrp) {
        return {};
    }

    data.reserve(((dec.data.size() - 1) * 5) / 8);
    if (ConvertBits<5, 8, false>([&](unsigned char c) { data.push_back(c); }, dec.data.begin(), dec.data.end())) {
        return data;
    }

    return {};
}
```
</p>
</details>